src/docs/arch-transaction.dox

```diff
@@ -1,8 +1,202 @@
+before calling another search. Read uncommitted transactions don't have a snapshot.
+
+If the transaction has a snapshot, each read will check whether the update's transaction id is
+in its snapshot. The updates with transaction ids in the snapshot or larger than the largest
+transaction id in the snapshot is not visible to the reading transaction.
+
+When operating in read committed or read uncommitted isolation levels, it is possible to read
+different values of the same key, seeing records not seen before, or finding records disappear
+in the same transaction. This is called phantom read. Under snapshot, WiredTiger guarantees
+repeated reads returning the same result except one case using timestamps.
+
+@subsection Timestamps
+
+WiredTiger provides a mechanism for the user to control when his operations should be visible,
+called timestamps. Timestamps are user specified sequence numbers that are associated with each
+operation. In addition, user can assign an immutable read timestamp to a transaction at the
+beginning. A transaction can only see updates with timestamps smaller or equal to its read
+timestamp. Note that read timestamp 0 means no read timestamp and the transaction can see the
+updates regardless of timestamps. Also note that timestamps don't have to be derived from
+physical times. User can use any 64 bit unsigned integer as logical timestamps. For a single
+operation, the timestamps associated with the operations in the same transaction don't have to
+be the same as long as they are monotonically increasing.
+
+Apart from the operation level timestamps, the users are also responsible for managing the
+global level timestamps, i.e, the oldest timestamp, and the stable timestamp. The oldest
+timestamp is the timestamp that should be visible by all concurrent transactions. The stable
+timestamp is the minimum timestamp that a new operation can commit at.
+
+Only transactions running in snapshot isolation can run with timestamps.
+
+@subsection Visibility
+
+The visibility of the transactions in WiredTiger considers both the operations' transaction ids
+and timestamps. The operation is visible only when both its transaction id and its timestamp
+are visible to the reading transaction.
+
+To read a key, WiredTiger first traverses all the updates of that key still in memory until a
+visible update is found. The in-memory updates in WiredTiger are organized as a singly linked
+list with the newest update at the head, called the update chain. If no value is visible on the
+update chain, it checks the version on the disk image, which is the version that was chosen to
+be written to disk in the last reconciliation. If it is still invisible, WiredTiger will search
+the history store to check if there is a version visible to the reader there.
+
+The repeated read guarantee under snapshot isolation may break in one case if the timestamps
+committed to the updates are out of order, e.g,
+
+U@20 -> U@30 -> U@15.
```

Comment:
Tip: To put this paragraph in monospace font, use three backticks - an illustration below with backtick (`) replaced by ~ (since github also treats backtick specially):
```
~~~
U@20 ....
~~~
```